site/docs/types/type_classes.md

```diff
@@ -31,16 +31,17 @@ Simple type classes are those that don't support any form of configuration. For
+| STRUCT&lt;T1,...,Tn&gt;      | A list of types in a defined order.                                                                                                                                                                                                                                                                                                                                   | `repeated Literal`, types matching T1..Tn
 | NSTRUCT&lt;N:T1,...,N:Tn&gt; | **Pseudo-type**: A struct that maps unique names to value types. Each name is a UTF-8-encoded string. Each value can have a distinct type. Note that NSTRUCT is actually a pseudo-type, because Substrait's core type system is based entirely on ordinal positions, not named fields. Nonetheless, when working with systems outside Substrait, names are important. | n/a
-| LIST&lt;T&gt;                | A list of values of type T. The list can be between [0..2,147,483,647] values in length. | `repeated Literal`, all types matching T
-| MAP&lt;K, V&gt;              | An unordered list of type K keys with type V values. Keys may be repeated. While the key type could be nullable, keys may not be null. | `repeated KeyValue` (in turn two `Literal`s), all key types matching K and all value types matching V
+| LIST&lt;T&gt;                | A list of values of type T. The list can be between [0..2,147,483,647] values in length.                                                                                                                                                                                                                                                                              | `repeated Literal`, all types matching T
+| MAP&lt;K, V&gt;              | An unordered list of type K keys with type V values. Keys may be repeated. While the key type could be nullable, keys may not be null.                                                                                                                                                                                                                                | `repeated KeyValue` (in turn two `Literal`s), all key types matching K and all value types matching V
+| TIMESTAMP&lt;P&gt;           | A naive timestamp within [1000-01-01 00:00:00.000000000..9999-12-31 23:59:59.999999999], with precision (P, number of digits) <= 9. Does not include timezone information and can thus not be unambiguously mapped to a moment on the timeline without context. Similar to naive datetime in Python.                                                                  | `int64` nanoseconds since 1970-01-01 00:00:00.000000000 (in an unspecified timezone)
```

Comment:
Probably should have both the old and new types listed here.

Reply:
From the proto it looks like we are deprecating the old type?  Are you saying we should list the old type and mark it deprecated?

Reply:
This does not work.  If you store int64 nanoseconds since the epoch then the maximum value you can get is `2262-04-12 01:47:16.854775807 +0200` (slightly larger if you store unsigned but still not large enough).  However, the first sentence claims `within [1000-01-01 00:00:00.000000000..9999-12-31 23:59:59.999999999]`.

So either the literal representation needs to be larger than 8 bytes or the allowable range needs to depend on the precision.

Note that making literals larger than 8 bytes is tempting, but it could lead to system incompatibility.  For example, you could create a Substrait filter `my_date < year_4000` but if you tried to evaluate that on the consumer it might try and convert that literal to its internal representation and return an error (e.g. this is what would happen if you tried to use pandas, datafusion, or pyarrow to evaluate that).

So I think we have to state that the allowable range depends on the precision.